### ISSUE TITLE
Support for using custom instance of axios

### ISSUE THREAD

**Issue by rchl** [ferdikoomen/openapi-typescript-codegen#942]
**Describe the solution you'd like**
When using the `httpClient: 'axios' option, I would like to be able to also use a custom axios instance to make the requests.

The reason for that is that I would like to use an Axios instance configured to my liking (custom interceptors, different timeout value, etc...).

Looking at the code it seems that it would be pretty easy to support that and I could contribute if the feature request is accepted.

**Comment by ferdikoomen** [comment]
How are you configuring Axios, globally or per instance? because you can already set global interceptors.

**Comment by rchl** [comment]
I'm creating a custom Axios instance and configuring it.

**Comment by smarlhens** [comment]
Hi there,

I may be interested in using a custom instance of Axios to override configuration of httpsAgent in order to get rid of the error `ERR_CERT_AUTHORITY_INVALID` when using auto-signed SSL certificates.

```typescript
const httpsAgent = new Agent({
  rejectUnauthorized: false,
});
```

Not sure this is the best way but I guess it would be nice if we could provide a custom instance.

PS: Apologize if this is not the right place to put that message, as the title is a little bit generic, I wanted to add my use case.

Wish you a nice day 🌞

**Comment by busheezy** [comment]
You can specify your axios interceptors and defaults before axios is used. There isn't a way for different settings for different named clients so this won't work for some use cases.

https://share.busheezy.dev/bOEKbJ.png

**Comment by IniterWorker** [comment]
I second @rchl, `axios` without a way to custom the current `axiosInstance` is not meaningful. But, I understand that the encapsulation HttpRequest doesn't want to carry custom behaviour from `axios`. So, what is the best approach to use `AxiosInstance` vs `axios` global?

**Comment by AllanPinheiroDeLima** [comment]
Having an option to use custom instances would be great. I actually have a problem where a can't use interceptors nor set language headers in a nuxt app because the framework encapsulates the options from the composables. 

Unless I'm doing something very wrong, I can't have multiple configs working for all parts of my app without dropping this lib entirely

**Comment by tibbe** [comment]
I ran into this as well. In addition to the general reasons for supporting a custom axios instance mentioned above (i.e. don't mutate global state that can be used by any part of the program), having a custom axios instance is required if you want to automate JWT token refresh (using interceptors) in case the current token has expired. You can't have the global instance put your credentials on all requests and not all request are hitting the same server (and thus need different credentials).

**Comment by sam-mfb** [comment]
Makes sense to me as well.  I think you also want the generated code to export the created axios instance so that your consumer code could set the default config for the instance via `myAxiosInstance.default` without messing with the generated code.

**Comment by Or-Zarchi-Forter** [comment]
I've create a PR that allows this, #1434. Please comment there so it will get attention 😄

**Comment by blipk** [comment]
https://github.com/ferdikoomen/openapi-typescript-codegen/blob/master/docs/custom-request-file.md

**Comment by dsuttonpreece** [comment]
> I've created a PR that allows this, #1434. Please comment there so it will get attention 😄

would love this PR to get merged!!

**Comment by unshame** [comment]
Since #1434 has been merged it's possible to do this. Although the way it's setup is awkward https://github.com/ferdikoomen/openapi-typescript-codegen/blob/master/docs/axios-support.md#using-a-custom-axios-client

I think wrapping it in a factory method makes it bearable:

```typescript
import { AxiosInstance } from 'axios'

import { ApiClient, CancelablePromise, OpenAPIConfig } from './generated'
import { ApiRequestOptions } from './generated/core/ApiRequestOptions'
import { request } from './generated/core/request'
import { AxiosHttpRequest } from './generated/core/AxiosHttpRequest'

export const createApiClient = (axiosInstance: AxiosInstance, options: Partial<OpenAPIConfig>) => {
  return new ApiClient(
    options,
    class AxiosHttpRequestInstance extends AxiosHttpRequest {
      public override request<T>(options: ApiRequestOptions): CancelablePromise<T> {
        return request(this.config, options, axiosInstance)
      }
    },
  )
}
```

**Comment by chamorim21** [comment]
> Since #1434 has been merged it's possible to do this. Although the way it's setup is awkward https://github.com/ferdikoomen/openapi-typescript-codegen/blob/master/docs/axios-support.md#using-a-custom-axios-client
> 
> I think wrapping it in a factory method makes it bearable:
> 
> ```ts
> import { AxiosInstance } from 'axios'
> 
> import { ApiClient, CancelablePromise, OpenAPIConfig } from './generated'
> import { ApiRequestOptions } from './generated/core/ApiRequestOptions'
> import { request } from './generated/core/request'
> import { AxiosHttpRequest } from './generated/core/AxiosHttpRequest'
> 
> export const createApiClient = (axiosInstance: AxiosInstance, options: Partial<OpenAPIConfig>) => {
>   return new ApiClient(
>     options,
>     class AxiosHttpRequestInstance extends AxiosHttpRequest {
>       public override request<T>(options: ApiRequestOptions): CancelablePromise<T> {
>         return request(this.config, options, axiosInstance)
>       }
>     },
>   )
> }
> ```

@unshame Where is `ApiClient` coming from? In my generated files there is no `ApiClient`. My services use directly the `_request` function.

**Comment by unshame** [comment]
You need to use the --name option to generate a class that you can instantiate.

**Comment by kimzeevaarders** [comment]
Can you not customize the axios request config over an interceptor? I had the exact same https agent issue and solved it with an request interceptor in the successor of this project. However, i'm not sure if it's possible to define (axios) interceptors in this project as well:

https://github.com/hey-api/openapi-ts/issues/512

**Comment by smarlhens** [comment]
Hey @kimzeevaarders 👋🏻 
Thanks for reaching out and for your solution 🙏🏻.
I apologize for the delayed response. Unfortunately, after all this time, I don't recall the specific solution I implemented to fix this issue (I'm not sure I succeeded 🤔).

**Comment by mrlubos** [comment]
Hey all, if you can migrate to [Hey API](https://heyapi.vercel.app/openapi-ts/migrating.html#openapi-typescript-codegen), this [client](https://heyapi.vercel.app/openapi-ts/clients.html) will be available soon 🚀
